Implemented IOS XR support for BGP address-family commands and properties and refactored tests to use getters/setters.

brturn opened the pull request.

Commit:
Implemented IOS XR support for BGP address-family commands and properties.

Commit:
Fixed skipping tests and rewrote long tests using getters/setters instead.

Commit:
Refactor test code to use getter/setters

glennmatthews commented:
Looks like you have a mismatch between your git user configuration and your github account - the commits don't seem to be associated with your 'brturn' github account.

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,21 +98,41 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
     ########################################################
 
+    # IOS XR?
+    def ios_xr?
```
I like this - consider moving it into the NodeUtil class so it can be shared to other provider classes.

RichWellum replied:
If you don't do this in this commit - can you add a bug to our backlog to track this please?

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -397,20 +446,24 @@ def dampening=(damp_array)
-        # Array not in a valid format
-        fail ArgumentError
+        # 'dampening route-map WORD' command
+        if platform == :nexus
```
Might as well create a `nexus?` method on NodeUtils as well.

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -502,12 +556,16 @@ def networks=(should_list)
         Cisco::Logger.debug("networks delta #{@get_args}\n #{action}: " \
-                            "#{delta_hash[action]}")
-        delta_hash[action].each do |network, route_map|
+                          "#{delta_hash[action]}")
```
This indentation change looks suspect.

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -533,7 +591,7 @@ def redistribute=(should)
       [:add, :remove].each do |action|
         Cisco::Logger.debug("redistribute delta #{@get_args}\n #{action}: " \
-                            "#{delta_hash[action]}")
+                          "#{delta_hash[action]}")
```
ditto comment about indentation. Yet rubocop is happy with it, it looks like - I'm confused. :-)

glennmatthews commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -1,44 +1,52 @@
   kind: boolean
   config_get_token_append: '/^advertise l2vpn evpn$/'
-  config_set_append: '<state> advertise l2vpn evpn'
+  config_set_append: ' <state> advertise l2vpn evpn'
```
Something's off here - if this command doesn't apply to XR, the leading space isn't needed. If it does apply, the _exclude is wrong. Same comment on dampen_igp_metric below.

brturn replied:
After Charles explained the bug to me, I did a replace-all.  I consider this a latent bug and if it doesn't hurt Nexus, then adding a space may possibly eliminate future copy-paste recurrences of the bug.

I've removed the space from all nexus-only lines, but left it in all XR and XR/NX lines.

glennmatthews commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -83,12 +104,28 @@ dampening_suppress_time:
+  cli_ios_xr:
+    # IOS XR does not have default-information under address-family,
+    #   so we must replace config get/set
+    # Perhaps better to fix in ruby code?
```
Yeah this looks sketchy. :-)

brturn replied:
Not sure what the right fix is?  Currently there is a check for vrf-unsupported features under XR, so I imagine this could be added as an address-family-unsupported feature?  But the other fix feels like a bit of a hack to begin with.  Any advice?

glennmatthews replied:
If this is a 'bgp_af' property that XR BGP doesn't support under any address families, then this property should be excluded on XR. (I guess on XR, from the comment, that this command is configured in some other scope? In that context it would then be appropriate to have a future work item to add a version of this property under the appropriate object for that scope.)

RichWellum replied:
FYI:

It's not supported under bgp af:

RP/0/RP0/CPU0:ios(config)#router bgp 5
RP/0/RP0/CPU0:ios(config-bgp)#address-family ipv4 unicast
RP/0/RP0/CPU0:ios(config-bgp-af)#bgp d?
dampening
RP/0/RP0/CPU0:ios(config-bgp-af)#bgp ?
  attribute-download  Configure attribute download for this address-family
  bestpath            Change default route selection criteria
  client-to-client    Configure client to client route reflection
  dampening           Enable route-flap dampening
  label-delay         Specify delay for batching label processing
  origin-as           BGP origin-AS knobs
  scan-time           Configure background scanner interval for this address-family

So _exclude here for XR, then add the logic minitest to catch the UnsupportedError. Unless it's a special case where for example it works under the af for a default vrf, but not a non-default.

robert-w-gries commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,21 +98,41 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
 
     def client_to_client=(state)
+      fail_vrf_unsupported_xr('client-to-client')
+      state = !state if ios_xr?
```
Can you add a comment that explains this line?  Not sure why it needs to be inversed.

robert-w-gries commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,21 +98,41 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
+      ios_xr? && @vrf != 'default'
+    end
+
+    def fail_vrf_unsupported_xr(reason)
```
Nit: rename method to `fail_if_vrf_unsupported_xr` so it's clear that the method does not always call the `fail` method

robert-w-gries commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -160,6 +180,32 @@ def default_next_hop_route_map
+
+    def next_hop_route_policy=(route_policy)
+      fail_vrf_unsupported_xr('nexthop route-policy')
+      route_policy.strip!
```
Adding `kind: string` to the yaml entry should eliminate the need for this line

Commit:
Modifications based on first-round comments from pull request.

Commit:
Fixing second round comments; exclude "default-information" on XR.

robert-w-gries commented on tests/test_bgp_af.rb:
```diff
@@ -314,65 +371,59 @@ def test_additional_paths_selection
     #
+    if platform == :ios_xr
+      config_ios_xr_dependencies(asn)
+      config('route-policy drop_all', '#drop_all', 'end-policy')
```
`'#drop_all'` seems like a no-op to me.  It was pulled in from #87, does anyone know the reason for that command being there?

Edit:  [This line](https://github.com/cisco/cisco-network-node-utils/pull/136/files#diff-09b95f5c8a820ab9f6db3a11831816deR289) has the same command

brturn replied:
Out of my area of expertise, can you suggest an alternative?  It looks like XR needs some additional setup before configuring the route-policy for additional-paths.  I think this is meant to be a mock policy?  Would it be better to setup a 'realistic' policy instead?

chburket replied:
I think it would probably be fine to just remove the '#drop_all' arg from the config call.  Not sure why Geert had it in there.

[edit] I just tried this on XR and it works fine without the "#drop_all"

RichWellum replied:
Why not move this into config_ios_xr_dependencies(asn)?

You can look at bgp neighbor af for similar:
  def setup_ios_xr
...
 # Global requirements
    # route policies
    str1 = 'route-map-in-name'
    str2 = 'route-map-out-name'
    config("route-policy #{str1}", 'end-policy')
    config("route-policy #{str2}", 'end-policy')
    config("route-policy #{str1.reverse}", 'end-policy')
    config("route-policy #{str2.reverse}", 'end-policy')
    config('route-policy foo_bar', 'end-policy')
    config('route-policy baz_inga', 'end-policy')

brturn commented:
@glennmatthews 

> Looks like you have a mismatch between your git user configuration and your github account
> the commits don't seem to be associated with your 'brturn' github account.

New to github; not sure what you are seeing as a problem?  It looks fine on my end.

glennmatthews commented:
Do you have the email address 'brturn@cisco.com' associated with your GitHub user account? GitHub doesn't appear to be recognizing that 'brturn@cisco.com' is GitHub user @brturn.

(Edit: looks to be fixed now - thanks!)

chburket commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -1,41 +1,49 @@
+  cli_ios_xr:
+    config_get: 'show running-config router bgp'
 
 additional_paths_install:
```
I think that additional_paths_install is not supported on XR at all (even under default vrf).  Can you double check?

brturn replied:
You are correct - I'll add that to be fixed in my next task.

chburket commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,21 +98,49 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
+      ios_xr? && @vrf != 'default'
+    end
+
+    def fail_if_vrf_unsupported_xr(reason)
```
Can you fix this so there are no extra spaces in the error message?

``` ruby
      fail Cisco::UnsupportedError.new(
        'bgp_af', reason, 'set',
        "#{reason} is not configurable on a "\
        'per-VRF basis on IOS XR') if vrf_unsupported_xr?
```

brturn replied:
Fixed

glennmatthews replied:
Might be more readable if you change 'reason' to 'property'? Nitpick.

glennmatthews commented:
:+1:

Commit:
Minor edits based on code review.

brturn commented:
@chburket @RichWellum @robert-w-gries Any last comments?  Thumbs up?

robert-w-gries commented:
:+1:

chburket commented:
:+1:

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -106,6 +110,41 @@ def test_collection_not_empty
+  def config_ios_xr_dependencies(asn, vrf='red', casty='unicast')
+    # These dependencies are required on ios xr
+
+    # router bgp 55
```
Nit. I think you can condense this example in the comments? 

And there's a couple of unnecessary comments in there like 'Configure'.

Also not sure what calling the variable 'casty' represents?

Is the loopback needed?

brturn replied:
Cleaned this up; loopback does not appear to be needed.

"casty" is just capturing the unicast/multicast of the address-family, it is only used in one place.  This will not be needed after migrating to a matrix-based test harness.

RichWellum replied:
Ack.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -118,114 +157,151 @@ def test_default_information_originate
+  ## next_hop route-map or route-policy
+  ##
+  def test_next_hop_route_map_or_policy
+    asn = '55'
```
You set the asn and vrf and af etc several times. Maybe a helper method? See bgp global for examples.

brturn replied:
Wasn't sure how much of the code to muck with for this story.  It does seem very redundant.

RichWellum replied:
This is really a nit, so if it's too much work then don't worry about it.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -269,22 +330,17 @@ def test_additional_paths
-      assert_match(pat, af_string,
-                   "Error: '#{pat}' is not configured and should be")
-    end
+    bgp_af.additional_paths_install = true if platform == :nexus
```
Should we set these for both platforms and then catch the error? Ditto to similar.

RichWellum commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -142,7 +170,7 @@ def next_hop_route_map
 
     def next_hop_route_map=(route_map)
-      route_map.strip!
+      #      route_map.strip!
```
Remove funky comment?

RichWellum commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -160,6 +188,32 @@ def default_next_hop_route_map
+
+    def next_hop_route_policy=(route_policy)
+      fail_if_vrf_unsupported_xr('nexthop route-policy')
+      #      route_policy.strip!
```
Ditto comment.

RichWellum commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -397,20 +454,24 @@ def dampening=(damp_array)
+          Cisco::Logger.debug("Dampening 'dampening #{route_map}'")
+        elsif ios_xr?
+          route_policy = "route-policy #{damp_array}"
+          #          route_policy.strip!
```
Comment ditto.

RichWellum commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -96,36 +119,60 @@ feature_nv_overlay_evpn:
+    config_get_token_append: '/^maximum-paths (\d+)$/'
+    config_set_append: '<state> maximum-paths <num>'
+  cli_ios_xr:
+    # BT: Is "maximum-paths" without specialization equal to "maximum-paths ebgp"?
```
Comment in the wild? Did you answer this question?

robert-w-gries replied:
ping @brturn

RichWellum commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -96,36 +119,60 @@ feature_nv_overlay_evpn:
+    config_set_append: '<state> network <network> <route_map>'
+  cli_ios_xr:
+    config_get_token_append: '/^network (\S+) ?(?:route-policy )?(\S+)?$/'
+    config_set_append: ' <state> network <network> <route_policy>'
```
The set append could be common to both right? I though that was the approach we are taking where possible. Applies globally.

RichWellum commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -96,36 +119,60 @@ feature_nv_overlay_evpn:
   default_value: []
 
 redistribute_policy:
   # route-map/policy is optional on some platforms, required on others
```
Don't think this comments adds much.

RichWellum commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -96,36 +119,60 @@ feature_nv_overlay_evpn:
+  cli_nexus:
+    config_set_append: '<state> redistribute <protocol> route-map <policy>'
+  cli_ios_xr:
+    config_set_append: ' <state> redistribute <protocol> route-policy <policy>'
```
I think these sets could be combined very easily. In the nu_setter - you'd format the string and pass that as the token. E.g instead of passing <policy> you'd pass <policy-string> equiv where it is route-map for nexus and route-policy for xr. We've done that in several places already.

brturn replied:
Didn't quite understand you here:  What is the "nu_setter"?  Are you suggesting passing "route-map" from the test harness when running on Nexus, and "route-policy" when running on XR, then just dropping that string into a platform-common command line here?

I was not so sure of the route-map/policy split, Geert mentioned that some versions of IOS may have both.  Code would be much cleaner if we collapsed them and dealt with the special case of having both at some later date.

RichWellum replied:
nu_setter : node_utils setter function.

No nothing to do with the test file; I am suggesting something like what I am doing in bgp af neighbor:

```
def default_originate_set(state, map=nil)
      if platform == :ios_xr
        map = "route-policy #{map}" unless map.nil?
      else
        map = "route-map #{map}" unless map.nil?
      end
      set_args_keys(state: (state ? '' : 'no'), map: map)
      config_set('bgp_neighbor_af', 'default_originate', @set_args)
    end
```

So map is then the token in the yaml:

```
default_originate:
  default_value: false
  cli_ios_xr:
    config_get_token_append: '/^default-originate(?: route-policy .*)?/'
    config_set_append: ' <state> default-originate <map>'
  cli_nexus:
    config_get_token_append: '/^default-originate(?: route-map .*)?/'
    config_set_append: '<state> default-originate <map>'
```

Of course I didn't make the set_append common which kind of makes this redundant, but I was expecting to clean that all up with the yaml User Story.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -28,7 +28,11 @@ def setup
-    config('no feature bgp', 'feature bgp')
+    if platform == :nexus
+      config('no feature bgp', 'feature bgp')
+    elsif platform == :ios_xr
```
Personally I think the elsif platform = xyz, could just be replaced with 'else' throughout the code. We only support two platforms currently. No biggy.

RichWellum replied:
This is a nit btw.

glennmatthews replied:
I'd rather leave it as is but am OK either way.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -314,65 +370,59 @@ def test_additional_paths_selection
     #
+    if platform == :ios_xr
+      config_ios_xr_dependencies(asn)
+      config('route-policy drop_all', 'end-policy')
```
Think I mentioned in a previous pr# this should go into the method above.

brturn replied:
Agreed.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -314,65 +370,59 @@ def test_additional_paths_selection
+
+    if platform == :ios_xr
+      assert_raises(Cisco::UnsupportedError) do
+        bgp_af.advertise_l2vpn_evpn = !initial
```
Do you need to do !initial? It's simply not supported regardless of what you set.

brturn replied:
I was told that setting it to the value it is already set at would cause no commands to occur.  I saw this happen in one of the tests and came up with the !initial as a general answer to the issue.  If I'm mistaken about the no-op, let me know!

RichWellum replied:
Right - I see the statement above now. Carry on. :)

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -390,123 +440,78 @@ def test_dampen_igp_metric
+               'dampening should return nil on XR with non-default vrf')
+    assert_raises(Cisco::UnsupportedError) do
+      bgp_af.dampening = %w(one two)
+    end
```
bgp_af.destroy ? Can you check you're doing this consistently for all test methods. Also in some you add an extra line before it and some you don't - can you make it consistent please?

brturn replied:
Agreed.

RichWellum commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -160,6 +188,32 @@ def default_next_hop_route_map
+        state = 'no'
+        # Dummy route policy required if not configured.
+        if next_hop_route_policy.empty?
+          route_policy = 'dummy_routepolicy'
```
I'd like to know if the team thinks that adding a dummy route_policy is ok, or is it better to fail (from the manifest) and add support for route-policy asap?

Good example is rd auto - we just failed until someone added it.

RichWellum commented on tests/test_bgp_af.rb:
```diff
@@ -118,114 +157,151 @@ def test_default_information_originate
+  ## client-to-client reflection (BT: trying a different style)
+  ##
+  def test_client_to_client
+    asn = '55'
```
Note that l2vpn is perfectly valid address-family:

router bgp 55
 address-family l2vpn evpn

You should add these tests throughout. You can see bgp neighbor af for examples on how to do this within a matrix that can be called for each test method. It's a harness I think should be adopted.

Ditto ipv6 and multicast.

robert-w-gries commented:
@brturn Can you please merge upstream `feature/grpc` branch into your fork, so you can pull in #145 and fix Travis?

Commit:
Rewrote most tests to use generic matrix-based property testing harness.
Should be able to capture several more tests also, probably left for next iteration.

Commit:
Minor edits

Commit:
Merge branch 'feature/grpc' of https://github.com/cisco/cisco-network-node-utils into feature/grpc
Minor merge edits.

Commit:
Corrected/skipped tests from upstream merge

To be fixed next iteration.

brturn commented:
Merged upstream - some new tests fixed, some skipped to get things compiling.
Rewrote test harness to use a matrix of tests and expected results.
@glennmatthews @RichWellum @robert-w-gries @chburket

robert-w-gries commented:
Rubocop is not passing according to Travis:

```
Offenses:

tests/test_bgp_af.rb:599:1: C: Trailing whitespace detected.
```

robert-w-gries commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -125,36 +149,62 @@ inject_map:
+  kind: string
   default_value: ""
+  config_get_token_append: '/nexthop route-policy (.*)$/'
+  config_set_append: ' <state> nexthop route-policy <route_policy>'
```
Shouldn't this property be removed since `next_hop_route_map` configures `route_policy` for XR?

Commit:
Minor edits for RuboCop, and removed nexthop route-policy specialization.

brturn commented:
Happy New Year!

Poking everyone about this PR again.  This is a major rewrite, collapsing most of the tests into a single harness.  We may need a sit-down code review?  I've left comments inline, and I'm sure it doesn't fit with typical Ruby code style. 

@glennmatthews @RichWellum @robert-w-gries @chburket

robert-w-gries commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
-    string
+  # rubocop:disable Style/WordArray
+  # rubocop:disable Style/ClassVars
+  # rubocop:disable Metrics/LineLength
```
- `Style/WordDisable` - Why do you want to disable this?  The matrices look cleaner to me with the word array:

``` diff
-    # rubocop:disable Style/WordArray
-    master = [['lax', 'sfo'],
-              ['lax', 'sjc'],
-              ['nyc', 'sfo', 'copy-attributes'],
-              ['sjc', 'nyc', 'copy-attributes']]
+    master = [%w(lax sfo),
+              %w(lax sjc),
+              %w(nyc sfo copy-attributes),
+              %w(sjc nyc copy-attributes)]
```
- `Style/ClassVars` - Do the test variables need to be class variables?  I feel this style doesn't need to be disabled.
- `Metrics/LineLength` - This doesn't need to be disabled.  Instead you can increase the `Max` value allowed, which is set inside `tests/.rubocop.yml`

PSA: WordArray is an autocorrectable offense, which means you can just run `rubocop -a` to change the arrays to word arrays.

brturn replied:
- Style/WordDisable -  I didn't write the inject_map code, so I don't know why the author disabled that one.  Looks about the same to me either way.
- ClassVars - I copied the matrix setup from test_bgp_neighbor_af.rb, which uses class vars.  When RuboCop complained I made them instance vars and then got lots of odd errors when running the tests.  Decided it wasn't worth the trouble and just added them back to being class vars.  Suggestions welcome.

glennmatthews replied:
Do you still need to disable the WordArray rule? Looks like you're using word arrays below now.

glennmatthews replied:
I would not favor changing Metrics/LineLength in .rubocop.yml - let's keep the current 80 character limit and disable on a case-by-case basis where it makes sense (like below)

glennmatthews replied:
It looks like the class variables are constants (they're defined once and never changed). Unless I'm missing something, try changing them to constants like `T_AFS`, `T_ASNS`, `T_VRFS`, etc. and see if the code works and rubocop is happier.

robert-w-gries replied:
@glennmatthews Good call regarding Metrics/LineLength.  I was confusing it with the Metrics/MethodLength cop

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -142,7 +170,6 @@ def next_hop_route_map
     end
 
     def next_hop_route_map=(route_map)
-      route_map.strip!
```
What was the reason behind removing these? Are they provably unnecessary?

robert-w-gries replied:
I suggested removing the `route_map.strip!` and adding `kind: string` to the YAML entry because `config_get` would automatically strip the white space.

glennmatthews replied:
Not sure I follow - this is on the setter side, not the getter side.

robert-w-gries replied:
Ah my mistake. @brturn can you add those two lines back in?

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -602,11 +635,15 @@ def networks=(should_list)
+            route_policy = "route-policy #{route_map_policy}"
+          end
+          set_args_keys(state: state, network: network, route_map: route_map,
+                        route_policy: route_policy)
```
The whole route-map vs route-policy behavior (where sometimes we treat them as the same thing and other times we treat them as completely separate) still smells wrong to me. Not blocking for this PR but I'll open a user story to revisit this in general.

RichWellum replied:
My initial idea was to delineate in the setter - create a key string where x = route-policy or route-map and pass x to the yaml. That was an idea from how we'd implemented something similar. The above seems odd to me too.

brturn replied:
I've been backing out the policy/map split slowly as I make edits.  I think Rich's proposal will work fine (just pass policy/map as a string to the setter).  I would prefer to push additional work to US59615.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
+    [:additional_paths_install,       [:toggle]],
+    [:advertise_l2vpn_evpn,           [:toggle]],
+
+    # TODO: To be added in next iteration
```
Do we have a user story for these TODOs?

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
+
+              # Getter should return nil when unsupported?
+              # BT: does not seem to work in general ..?
+              #            assert_nil(initial, "Assert 'nil' inital value failed for: #{test} #{asn} #{vrf} #{af}")
```
This should work - what value are you seeing for 'initial'?

Actually, you should maybe be testing `assert_equal(default, initial)` here just as you are doing in the 'else' case below.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -106,407 +329,55 @@ def test_collection_not_empty
-    ############################################
-    vrf = 'orange'
+    vrf = 'default'
+    vrf = 'orange' if platform == :nexus
```
Suggestion:

``` ruby
# this property is only applicable to the default vrf on XR
vrf = (platform == :ios_xr ? 'default' : 'orange')
```

Or alternately:

``` ruby
af = %w(ipv4 unicast)
if platform == :nexus
  vrf = 'orange'
elsif platform == :ios_xr
  vrf = 'default'
  config_ios_xr_dependencies(asn, vrf)
end
```

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -106,407 +329,55 @@ def test_collection_not_empty
     bgp_af.dampening = []
+    assert_equal(bgp_af.default_dampening,
+                 bgp_af.dampening) if platform == :nexus
+    assert_equal('', bgp_af.dampening) if platform == :ios_xr
```
is [] not the default on XR? if so then I'd prefer you update the `bgp_af.dampening = []` above since the comment indicates that we're setting to default.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -563,122 +404,79 @@ def test_dampening
-                 bgp_af.dampening_routemap,
-                 'The default dampening routemap should configured')
+    # IOS XR; we skip this section because we cannot query
+    #    the default parameters
```
This seems wrong to me - if this property is supported at all, it must have default values. Fix the provider instead of disabling the test?

brturn replied:
Charles an I looked at this one before I understood the default_values / auto_default semantics.  I think I could fix it now, but would prefer to push to US59615.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -745,6 +544,7 @@ def default_metric(asn, vrf, af)
   def test_default_metric
     afs = [%w(ipv4 unicast), %w(ipv6 unicast)]
     afs.each do |af|
+      config_ios_xr_dependencies('55') if platform == :ios_xr
```
Seems a little odd to pass `'55'` instead of `55` - is there a problem with the function definition?

brturn replied:
Passing the ASN as a string is very common in this file, not sure which is the preferred method?  Should it always be passed as a number instead?

glennmatthews replied:
Ideally both should work (string is sometimes needed to support ASDOT notation) but it's not critical especially for this PR. Just seemed inconsistent that this function takes `'55'` while the function below takes `55`. Not blocking.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -794,96 +594,17 @@ def inject_map_tester(bgp_af, should, test_id)
   end
 
   def test_inject_map
+    # Not sure why it's broken; to be fixed in next iteration
```
next iteration == another user story?

Please mark these with `# TODO:` to make them easier to find in future.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -1122,6 +879,7 @@ def test_utils_delta_add_remove
 
   # test route_target
   def test_route_target
+    skip('Currently broken on IOS XR: US59615') if platform == :ios_xr
```
please add a TODO comment here - thanks.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
+    %w(ipv6 multicast),
+    %w(l2vpn evpn),
+
+    # BT: These are additional address families/modifiers reported by XR, should they be tested also?
```
can you change all of these `BT:` to `TODO:`?

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
+            expect = check_test_exceptions(test, platform, vrf, af)
+
+            # Setup
+            bgp_af = RouterBgpAF.new(asn, vrf, af)
```
Just wondering - is it necessary to create/destroy the AF instance for each value under test? Or can this be moved outside the `@@tvalues.each` loop to improve performance of the test?

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -28,31 +28,254 @@ def setup
+              # This set of parameters produces a CLI error
+              assert_raises(Cisco::CliError,
+                            "Assert 'cli error' failed for: #{test}, #{asn}, #{vrf}, #{af}") do
+                bgp_af.send("#{test}=", 'foo') # *** BT: Don't like this ***
```
Agreed. Would it make more sense to try one (or all) of the values in `test_values` instead?

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -893,13 +614,21 @@ def test_network
+      end
+    end
     # Initial 'should' state
     if /ipv6/.match(dbg)
```
I know this is base code but since you're rewriting big chunks anyway - why are we regexping against the dbg string instead of checking `af.afi == 'ipv6'`?

brturn replied:
Would prefer to push new work to US59615.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -981,13 +724,22 @@ def redistribute_cmd(af, dbg)
     # rubocop:disable Style/WordArray
     # Initial 'should' state
     ospf = (dbg.include? 'ipv6') ? 'ospfv3 3' : 'ospf 3'
```
same question here about checking dbg instead of af.afi

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -981,13 +724,22 @@ def redistribute_cmd(af, dbg)
+                ['eigrp 1',   'my_policy'],
+                [ospf,        'my_policy'],
+                ['static',    'my_policy']]
+      master.push(['isis abc', 'my_policy']) if dbg.include? 'default'
```
`if af.vrf == 'default'`?

Commit:
Edits based on latest comments

glennmatthews commented:
I like the method_missing idea (though you also need to add `respond_to?`) but it's more than a bit of scope creep with this PR. :)

Can you provide latest minitest results against both XR and Nexus? If they look good, :+1: to merge this and move on.

brturn commented:
Minitest for Nexus and XR passes with some skips.

glennmatthews commented:
Are all of the skips explicitly triggered by your test code? (I.e., have you picked up the changes from #159 already)?

chburket commented:
I ran Bryan's latest code against my updated bgp_af beaker tests, and it passed on XR.  Running into a few issues on nexus involving route_target properties, but it looks like those are due to my changes, since the old beaker tests still run fine.

I'm good with merging and then fixing anything that pops up later.

:+1:

glennmatthews commented:
@brturn I'll merge as soon as I get a response on my last question - thanks for working through all this!

brturn commented:
I have not merged #159, do I need to rebase to the latest feature/grpc for that?

Currently, the skips come from two sources; an explicit 'skip' command in the test_bgp_af test function, or from executing a command for which _exclude: [cli_ios_xr] has been specified in the YAML.

glennmatthews commented:
With the change made in #159, only the first source (explicit `skip()`) is valid - the second case (trying to execute an unsupported feature) will be reported as an error and you need to update your test to catch it.

However, I'm going to go ahead and merge this PR as-is and you can fix any errored test cases as part of your next user story.
